File: Filewatcherd-Go/src/codewind/utils/pathutils.go
/*******************************************************************************
* Copyright (c) 2019 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

package utils

import (
	"codewind/models"
	"errors"
	"regexp"
	"runtime"
	"strings"
	"unicode"
)

func IsWindowsAbsolutePath(absolutePath string) bool {

	if len(absolutePath) < 2 {
		return false
	}

	char0 := absolutePath[0]

	if !unicode.IsLetter(rune(char0)) {
		return false
	}

	if absolutePath[1] != ':' {
		return false
	}

	return true

}

/** Ensure that the drive is lowercase for Unix-style paths from Windows. */
func NormalizeDriveLetter(absolutePath string) (string, error) {

	if strings.Contains(absolutePath, "\\") {
		return "", errors.New("This function does not support Windows-style paths")
	}

	if len(absolutePath) < 2 {
		return absolutePath, nil
	}

	if !strings.HasPrefix(absolutePath, "/") {
		return "", errors.New("Path should begin with forward slash: " + absolutePath)
	}

	char0 := absolutePath[0]
	char1 := absolutePath[1]

	// Special case the absolute path of only 2 characters.
	if len(absolutePath) == 2 {

		if char0 == '/' && unicode.IsLetter(rune(char1)) && unicode.IsUpper(rune(char1)) {
			return "/" + string(unicode.ToLower(rune(char1))), nil
		} else {
			return absolutePath, nil
		}
	}

	char2 := absolutePath[2]
	if char0 == '/' && char2 == '/' && unicode.IsLetter(rune(char1)) && unicode.IsUpper(rune(char1)) {

		return "/" + string(unicode.ToLower(rune(char1))) + string(char2) + absolutePath[3:], nil

	} else {
		return absolutePath, nil
	}
}

/** C:\helloThere -> /c/helloThere */
func ConvertFromWindowsDriveLetter(absolutePath string) string {

	if !IsWindowsAbsolutePath(absolutePath) {
		return absolutePath
	}

	absolutePath = strings.ReplaceAll(absolutePath, "\\", "/")

	char0 := absolutePath[0]

	// Strip first two characters
	absolutePath = absolutePath[2:]

	absolutePath = "/" + string(unicode.ToLower(rune(char0))) + absolutePath

	return absolutePath

}

/** Same as below, but determine behaviour based on OS. */
func ConvertAbsoluteUnixStyleNormalizedPathToLocalFile(str string) (string, error) {

	if runtime.GOOS != "windows" {
		// For Mac/Linux, nothing to do
		return str, nil
	}

	return ConvertAbsoluteUnixStyleNormalizedPathToLocalFileOS(str, true)
}

/* Convert /c/Users/Administrator to c:\Users\Administrator */
func ConvertAbsoluteUnixStyleNormalizedPathToLocalFileOS(str string, isWindows bool) (string, error) {

	if !isWindows {
		return str, nil
	}

	if !strings.HasPrefix(str, "/") {
		return "", errors.New("Parameters must begin with slash")
	}

	if len(str) <= 1 {
		return "", errors.New("Cannot convert string with length of 0 or 1: " + str)
	}

	driveLetter := str[1]

	if !unicode.IsLetter(rune(driveLetter)) {
		return "", errors.New("Missing drive letter: " + str)
	}

	if len(str) == 2 {
		return string(driveLetter) + ":\\", nil
	}

	secondSlash := str[2]
	if secondSlash != '/' {
		return "", errors.New("Invalid path format: " + str)
	}

	return string(driveLetter) + ":\\" + strings.ReplaceAll(str[3:], "/", "\\"), nil

}

type PathFilter struct {
	filenameExcludePatterns []*regexp.Regexp
	pathExcludePatterns     []*regexp.Regexp
}

func NewPathFilter(project *models.ProjectToWatch) (*PathFilter, error) {

	result := PathFilter{
		make([]*regexp.Regexp, 0),
		make([]*regexp.Regexp, 0),
	}

	ignoredFilenames := project.IgnoredFilenames
	if ignoredFilenames != nil {
		for _, val := range ignoredFilenames {

			if strings.Contains(val, "/") || strings.Contains(val, "\\") {
				return nil, errors.New("Ignore filenames may not contain path separators: " + val)
			}

			text := strings.ReplaceAll(val, "*", ".*")
			re, err := regexp.Compile(text)
			if err != nil {
				LogSevere("Unable to compile regex: " + text)
				return nil, err
			}

			result.filenameExcludePatterns = append(result.filenameExcludePatterns, re)

		}
	}

	ignoredPaths := project.IgnoredPaths
	if ignoredPaths != nil {
		for _, val := range ignoredPaths {
			if strings.Contains(val, "\\") {
				return nil, errors.New("Ignore paths may not contain Windows-style path separators: " + val)
			}

			text := strings.ReplaceAll(val, "*", ".*")
			re, err := regexp.Compile(text)
			if err != nil {
				LogSevere("Unable to compile regex: " + text)
				return nil, err
			}

			result.pathExcludePatterns = append(result.pathExcludePatterns, re)

		}
	}

	return &result, nil
}

func (p *PathFilter) IsFilteredOutByFilename(pathParam string) bool {

	if strings.Contains(pathParam, "\\") {
		LogSevere("Parameter cannot contain Window-style file paths")
		return false
	}

	strArr := strings.Split(pathParam, "/")

	// filename := path.Base(file)
	// if filename == "." {
	// 	return false
	// }

	for _, filename := range strArr {

		for _, val := range p.filenameExcludePatterns {
			if val.MatchString(filename) {
				return true
			}
		}

	}

	return false
}

func (p *PathFilter) IsFilteredOutByPath(path string) bool {

	if strings.Contains(path, "\\") {
		LogSevere("Parameter cannot contain Window-style file paths")
		return false
	}

	for _, val := range p.pathExcludePatterns {
		if val.MatchString(path) {
			return true
		}
	}

	return false

}

func ConvertAbsolutePathWithUnixSeparatorsToProjectRelativePath(path string, rootPath string) *string {

	// Strip project parent directory from path:
	// If pathToMonitor is: /home/user/codewind/project
	// and watchEventPath is: /home/user/codewind/project/some-file.txt
	// then this will convert watchEventPath to /some-file.txt

	if strings.Contains(path, "\\") {
		LogSevere("Parameter cannot contain Window-style file paths")
		return nil
	}

	rootPath = StripTrailingForwardSlash(rootPath)

	if !strings.HasPrefix(path, rootPath) {
		// This shouldn't happen, and is thus severe
		LogSevere("Watch event '" + path + "' does not match project path '" + rootPath + "'")
		return nil
	}

	path = strings.ReplaceAll(path, rootPath, "")

	if len(path) == 0 {
		// Ignore the empty case
		return nil
	}

	return &path

}

File: Filewatcherd-TypeScript/src/lib/PathUtils.ts
/*******************************************************************************
* Copyright (c) 2019 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

import * as log from "./Logger";

/**
 * Various utilities related to converting Windows-style paths (ex: c:\Users)
 * to/from our standardized Unix-style format (ex: /C/users), in order to
 * conform to our watcher API specification.
 */

/**
 * A windows absolute path will begin with a letter followed by a colon: C:\
 */
function isWindowsAbsolutePath(absolutePath: string): boolean {
    if (absolutePath.length < 2) {
        return false;
    }

    const char0 = absolutePath.charAt(0);

    if (!isLetter(char0)) {
        return false;
    }

    if (absolutePath.charAt(1) !== ":") {
        return false;
    }

    return true;
}

/** Ensure that the drive is lowercase for Unix-style paths from Windows. */
export function normalizeDriveLetter(absolutePath: string): string {

    if (absolutePath.indexOf("\\") !== -1) {
        throw new Error("This function does not support Windows-style paths: " + absolutePath);
    }
    if (absolutePath.length < 2) {
        return absolutePath;
    }

    if (!absolutePath.startsWith("/")) {
        throw new Error("Path should begin with forward slash: " + absolutePath);
    }

    const char0 = absolutePath.charAt(0);
    const char1 = absolutePath.charAt(1);

    // Special case the absolute path of only 2 characters.
    if (absolutePath.length === 2) {
        if (char0 === "/" && isLetter(char1) && isUpperCase(char1)) {
            return "/" + char1.toLowerCase();
        } else {
            return absolutePath;
        }

    }

    const char2 = absolutePath.charAt(2);
    if (char0 === "/" && char2 === "/" && isLetter(char1) && isUpperCase(char1)) {

        return "/" + char1.toLowerCase() + char2 + absolutePath.substring(3);

    } else {
        return absolutePath;
    }
}

/** C:\helloThere -> /c/helloThere */
export function convertFromWindowsDriveLetter(absolutePath: string): string {

    if (!isWindowsAbsolutePath(absolutePath)) {
        return absolutePath;
    }

    // Replace \ with /
    absolutePath = convertBackSlashesToForwardSlashes(absolutePath);

    const char0 = absolutePath.charAt(0);

    // Strip first two characters
    absolutePath = absolutePath.substring(2);

    absolutePath = "/" + char0.toLowerCase() + absolutePath;

    return absolutePath;

}

/* Convert /c/Users/Administrator to c:\Users\Administrator */
function convertAbsoluteUnixStyleNormalizedPathToLocalFileWithOS(str: string, isWindows: boolean): string {

    if (!isWindows) {
        return str;
    }

    if (!str.startsWith("/")) {
        throw new Error("Parameters must begin with slash");
    }

    if (str.length <= 1) {
        throw new Error("Cannot convert string with length of 0 or 1: " + str);
    }

    const driveLetter = str.charAt(1);
    if (!isLetter(driveLetter)) {
        throw new Error("Missing drive letter: " + str + " '" + driveLetter + "'");
    }

    if (str.length === 2) {
        return driveLetter + ":\\";
    }

    const secondSlash = str.charAt(2);
    if (secondSlash !== "/") {
        throw new Error("Invalid path format: " + str);
    }

    return driveLetter + ":\\" + convertForwardSlashesToBackSlashes(str.substring(3));

}

/** Same as below, but determine behaviour based on OS. */
export function convertAbsoluteUnixStyleNormalizedPathToLocalFile(str: string): string {

    if (!isOSWindows()) {
        // For Mac/Linux, nothing to do
        return str;
    }

    return convertAbsoluteUnixStyleNormalizedPathToLocalFileWithOS(str, true);
}

function convertBackSlashesToForwardSlashes(str: string): string {
    return str.split("\\").join("/");
}

function convertForwardSlashesToBackSlashes(str: string): string {
    return str.split("/").join("\\");
}

function isOSWindows() {
    return process.platform === "win32";
}

function isLetter(currentChar: string) {
    return ("a" <= currentChar && currentChar <= "z")
        || ("A" <= currentChar && currentChar <= "Z");
}

function isUpperCase(currentChar: string) {
    return ("A" <= currentChar && currentChar <= "Z");
}

export function stripTrailingSlash(str: string): string {

    while (str.trim().endsWith("/")) {

        str = str.trim();
        str = str.substring(0, str.length - 1);

    }

    return str;

}

// Strip project parent directory from path:
// If pathToMonitor is: /home/user/codewind/project
// and watchEventPath is: /home/user/codewind/project/some-file.txt
// then this will convert watchEventPath to /some-file.txt
export function convertAbsolutePathWithUnixSeparatorsToProjectRelativePath(path: string, rootPath: string): string  {

    if (rootPath.indexOf("\\") !== -1) {
        throw new Error("Forward slashes are not supported.");
    }

    rootPath = stripTrailingSlash(rootPath);

    if (!path.startsWith(rootPath)) {
        // This shouldn't happen, and is thus severe
        log.severe("Watch event '" + path + "' does not match project path '" + rootPath + "'");
        return null;
    }

    path = path.replace(rootPath, "");

    if (path.length === 0) {
        // Ignore the empty case

        return null;
    }

    return path;

}

/**  C:\helloThere -> /c/helloThere */
export function normalizePath(pathParam: string): string {

    // Convert \ to /
    let absPath = pathParam.split("\\").join("/");

    absPath = convertFromWindowsDriveLetter(absPath);
    absPath = normalizeDriveLetter(absPath);

    return absPath;

}

/** "/moo/cow" => [ "/moo/cow", "/moo"] */
export function splitRelativeProjectPathIntoComponentPaths(path: string): string[] {
    const result: string[] = [];

    let currPath = path;
    while (true) {

        if (currPath.length === 1) {
            break;
        }

        result.push(currPath);

        const index = currPath.lastIndexOf("/");
        if (index <= 0) {
            break;
        }

        currPath = currPath.substring(0, index);
    }

    return result;
}
